packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.25;
+
+import { Encoding } from "src/libraries/Encoding.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+import { TransientContext, TransientReentrancyAware } from "src/libraries/TransientContext.sol";
+import { CrossL2Inbox } from "src/L2/CrossL2Inbox.sol";
+import { IL2ToL2CrossDomainMessenger } from "src/L2/IL2ToL2CrossDomainMessenger.sol";
+import { ISemver } from "src/universal/ISemver.sol";
+
+/// @notice Thrown when a non-written slot in transient storage is attempted to be read from.
+error NotEntered();
+
+/// @notice Thrown when attempting to send a message to the chain that the message is being sent from.
+error MessageDestinationSameChain();
+
+/// @notice Thrown when attempting to relay a message and the function caller (msg.sender) is not CrossL2Inbox.
+error RelayMessageCallerNotCrossL2Inbox();
+
+/// @notice Thrown when attempting to relay a message where CrossL2Inbox's origin is not L2ToL2CrossDomainMessenger.
+error CrossL2InboxOriginNotL2ToL2CrossDomainMessenger();
+
+/// @notice Thrown when attempting to relay a message whose destination chain is not the chain relaying it.
+error MessageDestinationNotRelayChain();
+
+/// @notice Thrown when attempting to relay a message whose target is CrossL2Inbox.
+error MessageTargetCrossL2Inbox();
+
+/// @notice Thrown when attempting to relay a message whose target is L2ToL2CrossDomainMessenger.
+error MessageTargetL2ToL2CrossDomainMessenger();
+
+/// @notice Thrown when attempting to relay a message that has already been relayed.
+error MessageAlreadyRelayed();
+
+/// @custom:proxied
+/// @custom:predeploy 0x4200000000000000000000000000000000000023
+/// @title L2ToL2CrossDomainMessenger
+/// @notice The L2ToL2CrossDomainMessenger is a higher level abstraction on top of the CrossL2Inbox that provides
+///         features necessary for secure transfers ERC20 tokens between L2 chains. Messages sent through the
+///         L2ToL2CrossDomainMessenger on the source chain receive both replay protection as well as domain binding.
+contract L2ToL2CrossDomainMessenger is IL2ToL2CrossDomainMessenger, ISemver, TransientReentrancyAware {
+    /// @notice Storage slot for the sender of the current cross domain message.
+    ///         Equal to bytes32(uint256(keccak256("l2tol2crossdomainmessenger.sender")) - 1)
+    bytes32 internal constant CROSS_DOMAIN_MESSAGE_SENDER_SLOT =
+        0xb83444d07072b122e2e72a669ce32857d892345c19856f4e7142d06a167ab3f3;
+
+    /// @notice Storage slot for the source of the current cross domain message.
+    ///         Equal to bytes32(uint256(keccak256("l2tol2crossdomainmessenger.source")) - 1)
+    bytes32 internal constant CROSS_DOMAIN_MESSAGE_SOURCE_SLOT =
+        0x711dfa3259c842fffc17d6e1f1e0fc5927756133a2345ca56b4cb8178589fee7;
+
+    /// @notice Current message version identifier.
+    uint16 public constant MESSAGE_VERSION = uint16(0);
```

Comment:
small nit: this breaks the style guide, it should be `messageVersion` - our other contracts are offenders of this

Reply:
good idea! will fix now

Reply:
interestingly, it appears like `L2ToL1MessagePasser` and `CrossDomainMessenger` use `MESSAGE_VERSION`. couldn't find instances of `messageVersion`. what should we use?